adapters/repos/db/roaringsetrange/segment_cursor.go

```diff
@@ -53,3 +63,153 @@ func (c *SegmentCursor) Next() (uint8, roaringset.BitmapLayer, bool) {
 		Deletions: sn.Deletions(),
 	}, true
 }
+
+// ================================================================================
+
+// A SegmentCursor iterates over all key-value pairs in a single disk segment.
+// You can start at the beginning using [*SegmentCursorPread.First] and move forward
+// using [*SegmentCursorPread.Next]
+type SegmentCursorPread struct {
+	readSeeker     io.ReadSeeker
+	reader         *bufio.Reader
+	lenBuf         []byte
+	nodeBufs       [][]byte
+	nodeBufPos     int
+	nodeBufMinSize int
+}
+
+// NewSegmentCursorPread creates a cursor for a single disk segment. Make sure that
+// the reader has offset = 0 set correctly to start at the payload, as calling
+// [*SegmentCursorPread.First] will start reading at offset 0.
+// Similarly, the reader may only read payload, as the EOF
+// is used to determine if more keys can be found.
+//
+// bufferCount tells how many exlusive payload buffers should be used to return
+// expected data. Set multiple buffers if data returned by First/Next will not be used
+// before following call will be made, not to overwrite previously fetched values.
+// (e.g. count 3 means, 3 buffers will be used internally and following calls to First/Next
+// will return data in buffers: 1, 2, 3, 1, 2, 3, ...)
+func NewSegmentCursorPread(readSeeker io.ReadSeeker, bufferCount int) *SegmentCursorPread {
+	readSeeker.Seek(0, io.SeekStart)
+	return &SegmentCursorPread{
+		readSeeker:     readSeeker,
+		reader:         bufio.NewReaderSize(readSeeker, 10*1024*1024),
+		lenBuf:         make([]byte, 8),
+		nodeBufs:       make([][]byte, bufferCount),
+		nodeBufPos:     0,
+		nodeBufMinSize: 0,
+	}
+}
+
+func (c *SegmentCursorPread) First() (uint8, roaringset.BitmapLayer, bool) {
+	c.readSeeker.Seek(0, io.SeekStart)
+	c.reader.Reset(c.readSeeker)
+	return c.read(true)
+}
+
+func (c *SegmentCursorPread) Next() (uint8, roaringset.BitmapLayer, bool) {
+	return c.read(false)
+}
+
+func (c *SegmentCursorPread) read(isFirst bool) (uint8, roaringset.BitmapLayer, bool) {
+	_, err := io.ReadFull(c.reader, c.lenBuf)
+	if err == io.EOF {
+		return 0, roaringset.BitmapLayer{}, false
+	}
+	if err != nil {
+		panic(fmt.Errorf("SegmentCursorReader::Next reading node length: %w", err))
+	}
+
+	nodeLen := binary.LittleEndian.Uint64(c.lenBuf)
+	nodeBuf := c.getNodeBuf(int(nodeLen))
+
+	_, err = io.ReadFull(c.reader, nodeBuf[8:])
+	if err != nil {
+		panic(fmt.Errorf("SegmentCursorReader::Next reading node: %w", err))
```

Comment:
erroring would be preferred over panicking whenever possible

Reply:
Agreed. Even had errors in cursor's method signatures, but api and error handling got so messy, I've removed it in favour of completing implementation. Can try again with errors in separate PR